File: src/main/java/com/now/nowbot/service/MessageServiceImpl/OverSRService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.throwable.ServiceException.OverSRException;
import org.springframework.stereotype.Service;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service("OVERSR")
public class OverSRService implements MessageService<Matcher> {

    Pattern pattern = Pattern.compile("^[!！]\\s*(?i)(ym)?((oversr)|or(?![a-zA-Z_]))+(\\s+(?<SR>[0-9.]*))?");

    @Override
    public boolean isHandle(MessageEvent event, DataValue<Matcher> data) {
        var m = pattern.matcher(event.getRawMessage().trim());
        if (m.find()) {
            data.setValue(m);
            return true;
        } else return false;
    }

    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        String SRStr = matcher.group("SR");
        double SR;
        String message;

        try {
            SR = Double.parseDouble(SRStr);
        } catch (Exception e) {
            if (e instanceof NullPointerException) throw new OverSRException(OverSRException.Type.OV_Parameter_Null);
            else throw new OverSRException(OverSRException.Type.OV_Parameter_Error);
        }

        message = OverSR(SR);
        event.getSubject().sendMessage(message);
    }

    private String OverSR(double SR) throws OverSRException {
        StringBuilder message = new StringBuilder();
        double minute;
        double hour;
        double day;

        if ((int) (SR * 100f) <= 570) {
            throw new OverSRException(OverSRException.Type.OV_Send_Success);
        } else if ((int) (SR * 100f) <= 2000){
            // 超 0.01 星加 10 分钟，6星 以上所有乘以二
            if ((int) (SR * 100f) <= 600) {
                minute = (SR - 5.7f) * 1000f;
            } else {
                minute = (SR - 5.7f) * 2000f;
            }
        } else {
            throw new OverSRException(OverSRException.Type.OV_Parameter_OutOfRange);
        }

        message.append("已超星，预计禁言：");

        if (minute >= 1440f) {
            day = Math.floor(minute / 1440f);
            hour = Math.floor((minute - (day * 1440f)) / 60f);
            minute = Math.floor(minute - (day * 1440f) - (hour * 60f));

            message.append(String.format("%d", (int) day))
                    .append("天")
                    .append(String.format("%d", (int) hour))
                    .append("时")
                    .append(String.format("%d", (int) minute))
                    .append("分");
        } else if (minute >= 60f) {
            hour = Math.floor(minute / 60f);
            minute = Math.floor(minute - (hour * 60f));

            message.append(String.format("%d", (int) hour))
                    .append("时")
                    .append(String.format("%d", (int) minute))
                    .append("分");
        } else {
            message.append(String.format("%d", (int) minute))
                    .append("分");
        }

        return message.toString();
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/OverSRService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.NewbieRestrictService
import com.now.nowbot.service.NewbieRestrictService.Companion.STAR_BOUNDARY
import com.now.nowbot.throwable.botRuntimeException.IllegalArgumentException
import com.now.nowbot.util.Instruction
import com.now.nowbot.util.command.REG_NUMBER_DECIMAL
import org.springframework.stereotype.Service
import kotlin.math.roundToLong

@Service("OVER_SR") class OverSRService : MessageService<Double> {
    override fun isHandle(event: MessageEvent, messageText: String, data: DataValue<Double>): Boolean {
        val m = Instruction.OVER_SR.matcher(messageText)
        if (m.find()) {
            val starStr: String? = m.group("SR")

            if (starStr.isNullOrBlank() || starStr.matches(REG_NUMBER_DECIMAL.toRegex()).not()) {
                throw IllegalArgumentException.WrongException.Henan()
            }

            val star = starStr.toDoubleOrNull()
                ?: throw IllegalArgumentException.WrongException.StarRating()

            data.value = star
            return true
        } else return false
    }

    @Throws(Throwable::class) override fun handleMessage(event: MessageEvent, param: Double): ServiceCallStatistic? {
        event.replyAsync(getMessage(param))
        return ServiceCallStatistic.building(event) {
            setParam(mapOf(
                "star" to param.toFloat().toString()
            ))
        }
    }

     private fun getMessage(star: Double): String {
        val message = StringBuilder()
         val silence: Long

        if (star < STAR_BOUNDARY) {
            throw IllegalArgumentException("未超星。")
        } else if (star < 20.0) {
            silence = ((star - STAR_BOUNDARY) * 2000).roundToLong()
        } else {
            throw IllegalArgumentException.ExceedException.StarRating()
        }

        message.append("已超星，预计禁言：")
         message.append(NewbieRestrictService.getTime(silence))

        return message.toString()
    }
}
